### PR TITLE
Some arith and IrBuilder cleanup

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ir/builder.cpp [hunk 1]
```diff
@@ -271,13 +267,14 @@ Val* SimplifyingIrBuilder::addExpr(Val* lhs, PolymorphicValue rhs) {
   } else if (lhs == nullptr) {
     return IrBuilder::IrBuilder::create<Val>(rhs);
   }
-  auto target_dtype = promoteType(lhs->dtype(), getDataType(rhs));
```
Comment (naoyam) [inline csrc/ir/builder.cpp:274]: Why is this removed?

Comment (zasdfgbnm) [reply csrc/ir/builder.cpp:274]: I thought it didn't make sense, so I removed it. Because we can not accurately recover dtype from `PolymorphicValue`, for example, if `rhs` is actually int32, `getDataType(rhs)` will be int64, and if `lhs` is also int32, we might incorrectly promote the dtype to int64.

But after a second thought, I am feeling that we should better keep it rather than remove it. What if `rhs` is double and `lhs` is int64? For this case, having a result dtype double is definitely more correct than int64.

Comment (naoyam) [reply csrc/ir/builder.cpp:274]: > for example, if rhs is actually int32, getDataType(rhs) will be int64

When does `getDataType(rhs)` return int64 if it's actually int32?

Comment (zasdfgbnm) [reply csrc/ir/builder.cpp:274]: Recall that `PolymorphicValue` can only be `int64_t`, there is no `int` variant. So `getDataType(PolymorphicValue)` will always return int64 if the `PolymorphicValue` is holding an integer.

Comment (zasdfgbnm) [reply csrc/ir/builder.cpp:274]: Added back type promotion. Also added an optional parameter allowing users to specify the actual rhs dtype.
